### PR TITLE
ref(metrics-indexer): Indexer shared strings

### PR DESCRIPTION
**Pull request by MeredithAnya**
**context**
The indexer schema changed in https://github.com/getsentry/sentry/pull/32540 to add an `organization_id` (among other things). Although most strings are still specific to an organization, there are common tag keys and values, as well as metric names that will be used across organizations. We want to avoid storing a new record every time a new org uses one of these shared strings. 

**options**
1. We can use `organization_id=0` to store these strings once in the DB. We just need to make sure that we convert the org ids before we look values up in the cache/db. 

2. We can hard code the list of strings to ids in the code. (this is the chosen way)

note: relevant conversations in https://github.com/getsentry/sentry/pull/33279

Both these options require reserving ids for these shared strings (whether or not we actually store the records in the db) and that was handled in https://github.com/getsentry/sentry/pull/33321. We are going with option 2.

### CODE REVIEW

File: src/sentry/sentry_metrics/indexer/postgres_v2.py
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:95]: This method is not merging `KeyResults` instances nor it uses any of the `KeyResult` functionalities. So it does not need to be in this class. Its responsibility is separate. I would suggest you just make it a stand alone function.

File: src/sentry/sentry_metrics/indexer/postgres_v2.py
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:106]: I think this does not work.
What if the org id is not yet present in `self.results` ?

File: src/sentry/sentry_metrics/indexer/postgres_v2.py
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:321]: rename: `StaticStringsIndexerDecorator`

File: src/sentry/sentry_metrics/indexer/postgres_v2.py
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/postgres_v2.py:330]: I don't think you need this method. It is used only once and it is a one liner.

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (fpacifici) [inline src/sentry/sentry_metrics/indexer/strings.py:1]: Please explain why, what they are used for and why, if that happens it is a catastrophe.
Also You can add a simple unit test that checks that the number of strings is the expected one.
So if somebody adds a string, the test will fail and they will adjust the test, annoying but not too bad.
If somebody messes up this and removes a string, the test fails when it should not.

Basically it gives you a safeguard when this changes to ensure your change is deliberate when you make one.

File: src/sentry/sentry_metrics/multiprocess.py
Comment (fpacifici) [inline src/sentry/sentry_metrics/multiprocess.py:404]: How was this working before ?

File: src/sentry/sentry_metrics/indexer/postgres_v2.py
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/postgres_v2.py:106]: `self.results` is a `defaultdict(dict)` so it should just be updating an empty dict iirc

File: src/sentry/sentry_metrics/multiprocess.py
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/multiprocess.py:404]: what do you mean? we weren't using the new bulk_record before this change

File: tests/sentry/sentry_metrics/test_postgres_indexer.py
Comment (jjbayer) [inline tests/sentry/sentry_metrics/test_postgres_indexer.py:47]: Does the test require these entries in the database? `bulk_record` gets them from the static dict, right?

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (jjbayer) [inline src/sentry/sentry_metrics/indexer/strings.py:7]: It might not matter technically, but should we leave some space between the last metric name (4) and the first tag (5), just in case we get more shared metric names in the future? Or would it be a better idea to reserve ranges per use case, e.g. `0..99` for sessions, `100..199` for transactions?

File: tests/sentry/sentry_metrics/test_postgres_indexer.py
Comment (MeredithAnya) [reply tests/sentry/sentry_metrics/test_postgres_indexer.py:47]: ah yeah good point, had copied this over from before but needs to be updated

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/strings.py:7]: yeah was thinking about this too, seems easy enough to split into ranges

File: src/sentry/snuba/metrics/datasource.py
Comment (MeredithAnya) [inline src/sentry/snuba/metrics/datasource.py:193]: `-1` is not a valid metric id or tag id to send to ClickHouse

File: src/sentry/testutils/cases.py
Comment (MeredithAnya) [inline src/sentry/testutils/cases.py:1237]: We should not be sending `None` tags to snuba - we can create the record if it doesn't already exist

File: tests/sentry/api/endpoints/test_organization_metric_tag_details.py
Comment (MeredithAnya) [inline tests/sentry/api/endpoints/test_organization_metric_tag_details.py:159]: `release` is in the indexer so use random tag instead of random metric name

File: tests/sentry/search/events/test_builder.py
Comment (MeredithAnya) [inline tests/sentry/search/events/test_builder.py:1211]: should `transaction` ,`transaction.status`, `internal_error`, `ok` all be in `SHARED_TAG_STRINGS`? cc @jan-auer

**Review by fpacifici — COMMENTED** [review]
Looks good to me.
Please fix the last failing test, then you may want a quick review by somebody who has context on the product test cases you made changes to.

File: src/sentry/snuba/metrics/datasource.py
Comment (jjbayer) [reply src/sentry/snuba/metrics/datasource.py:193]: -1 is returned by the `resolve_weak` helper, see https://github.com/getsentry/sentry/blob/e063ea75be554e5e4a237d619645d7924a304074/src/sentry/sentry_metrics/utils.py#L6-L7

@ahmedetefy should `generate_metric_ids` call `indexer.resolve` instead of `resolve_weak`, so that we are consistent with line 182?

File: src/sentry/testutils/cases.py
Comment (jjbayer) [reply src/sentry/testutils/cases.py:1236]: Should be sufficient to replace `indexer.resolve` with `indexer.record` though, right? Since `record` basically does "resolve or record".

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (ahmedetefy) [inline src/sentry/sentry_metrics/indexer/strings.py:51]: For release health, this is missing a `errored_preaggr` status

Comment (ahmedetefy) [inline src/sentry/sentry_metrics/indexer/strings.py:51]: I noticed that transaction tags were not added, is that by design ?
If not, then here are the ones we need to add
```suggestion
    "transaction.status",
    "satisfaction",
    "ok",
    "cancelled",
    "unknown",
    "aborted",
    "satisfied"
    "tolerated"
    "frustrated"
}
```

File: src/sentry/snuba/metrics/datasource.py
Comment (ahmedetefy) [reply src/sentry/snuba/metrics/datasource.py:193]: Yes I agree with that
@MeredithAnya I will put a PR after you merge yours to incorporate Joris' suggestion

File: tests/sentry/search/events/test_builder.py
Comment (ahmedetefy) [reply tests/sentry/search/events/test_builder.py:1211]: I spoke with @jan-auer about this and he thinks we should go ahead with this.. I posted all the tag names/values for transactions in an earlier comment

File: src/sentry/testutils/cases.py
Comment (MeredithAnya) [reply src/sentry/testutils/cases.py:1236]: yes good point, will update

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/strings.py:91]: added

**Review by fpacifici — APPROVED** [review]
I think there is still a test to fix. Please fix that before merging.

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (ahmedetefy) [inline src/sentry/sentry_metrics/indexer/strings.py:1]: These strings look good to me!

**Review by ahmedetefy — APPROVED** [review]

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (jjbayer) [inline src/sentry/sentry_metrics/indexer/strings.py:27]: Does this mean that the range 0..65536 is now unused? Not saying it's a problem, just trying to understand.

File: src/sentry/sentry_metrics/indexer/strings.py
Comment (jjbayer) [inline src/sentry/sentry_metrics/indexer/strings.py:90]: Let's add `transaction.op`, `http.method`, `browser.name` and `os.name` here. They are not extracted yet, but will be in the future.

Comment (jjbayer) [reply src/sentry/sentry_metrics/indexer/strings.py:90]: @untitaker are the tag key and tag values for outlier tagging already defined? If so, we should add them here too.

Comment (untitaker) [reply src/sentry/sentry_metrics/indexer/strings.py:90]: I didn't think about it yet. On a whim I would say we can add:

* `histogram_outlier` -- this is a tag name
* `outlier` -- tag value
* `inlier` -- tag value

question: if we decide later that we are not happy with the naming of the string, is it easy to rename? since they are now technically defined in code only

Comment (MeredithAnya) [reply src/sentry/sentry_metrics/indexer/strings.py:90]: @untitaker it will be easy to rename if the name that we want instead is not already taken (aka someone sent us data and then we recorded it in the indexer).

**Pull request merged**
